From 1928b62c07a4acc3b4f942d35ebba65ba429064c Mon Sep 17 00:00:00 2001
From: t <tristan75015@gmail.com>
Date: Wed, 7 Oct 2026 03:26:02 +0200
Subject: [PATCH] MFront: type of the tangent operator checked before its
 conversion

The functions converting the tangent operator of a finite strain
behaviour read the operator computed by the code of the user with the
`get` method of `TrivialGenTypeBase`, which only checks its type on the
host. If this code does not compute the expected operator, the host throws
a `GenTypeCastError` exception, but device code read a value of another
type and reported a success. The type is now checked before the
conversion: the same exception is thrown on the host and the computation
fails in device code.
---
 mfront/src/BehaviourCodeGeneratorBase.cxx | 20 ++++++++++++++++++--
 1 file changed, 18 insertions(+), 2 deletions(-)

diff --git a/mfront/src/BehaviourCodeGeneratorBase.cxx b/mfront/src/BehaviourCodeGeneratorBase.cxx
index 78ac02ade..679036ef3 100644
--- a/mfront/src/BehaviourCodeGeneratorBase.cxx
+++ b/mfront/src/BehaviourCodeGeneratorBase.cxx
@@ -4788,7 +4788,15 @@ namespace mfront {
                 os << "if(this->computePredictionOperator_" << k
                    << "(smt) != SUCCESS){\n"
                    << "return FAILURE;\n"
-                   << "}\n"
+                   << "}\n";
+                // the type of the operator computed by the user is only
+                // checked by the `get` method on the host
+                os << "if(!this->Dt.template is<" << from_type_flag
+                   << "<N,stress>>()){\n";
+                this->writeHostAndDeviceCode(
+                    os, "tfel::raise<tfel::utilities::GenTypeCastError>();",
+                    "return FAILURE;");
+                os << "}\n"
                    << "const " << from_type_flag << "<N,stress> "
                    << "tangentOperator_" << k << " = this->Dt.template get<"
                    << from_type_flag << "<N,stress> >();\n";
@@ -4941,7 +4949,15 @@ namespace mfront {
                 os << "if(!this->computeConsistentTangentOperator_" << k
                    << "(smt)){\n"
                    << "return false;\n"
-                   << "}\n"
+                   << "}\n";
+                // the type of the operator computed by the user is only
+                // checked by the `get` method on the host
+                os << "if(!this->Dt.template is<" << k_type_flag
+                   << "<N,stress>>()){\n";
+                this->writeHostAndDeviceCode(
+                    os, "tfel::raise<tfel::utilities::GenTypeCastError>();",
+                    "return false;");
+                os << "}\n"
                    << "const " << k_type_flag << "<N,stress> "
                    << "tangentOperator_" << k << " = this->Dt.template get<"
                    << k_type_flag << "<N,stress>>();\n";
-- 
2.43.0

